probe_src/probe_frontend/lib/src/transcribe.rs

```diff
@@ -0,0 +1,419 @@
+/// 5. [`OpsArena`]s are parsed into which are then parsed into [`ops::Op`]s using the
+///    [`ArenaContext`].
+/// 6. [`ops::Op`]s are serialized into json and written line-by-line into the output directory.
+///
+/// (steps 5 & 6 are done lazily with iterators to reduce unnecessary memory allocations)
+///
+/// On success, returns the number of Ops processed in the TID directory.
+pub fn parse_tid<P1: AsRef<Path>, P2: AsRef<Path>>(in_dir: P1, out_dir: P2) -> Result<usize> {
+    fn try_files_from_dir<P: AsRef<Path>>(dir: P) -> Result<Vec<PathBuf>> {
+        match fs::read_dir(&dir) {
+            Ok(entry_iter) => entry_iter
+                .map(|entry_result| {
+                    entry_result
+                        .map(|entry| entry.path())
+                        .wrap_err("Error reading DirEntry from record TID subdirectory")
+                })
+                .collect::<Result<Vec<_>>>(),
+            Err(e) => Err(e.convert("Error opening record TID directory")),
+        }
+    }
+
+    // STEP 1
+    let tid = filename_numeric(&in_dir)?;
+    let mut outfile = {
+        let mut path = out_dir.as_ref().to_owned();
+        path.push(tid.to_string());
+        File::create_new(path).wrap_err("Failed to create TID output file")?
+    };
+
+    // STEP 2
+    let paths = fs::read_dir(&in_dir)
+        .wrap_err("Error reading record TID directory")?
+        .filter_map(|entry_result| match entry_result {
+            Ok(entry) => Some((entry.file_name(), entry)),
+            Err(e) => {
+                log::warn!("Error reading DirEntry in TID directory: {}", e);
+                None
+            }
+        })
+        .collect::<HashMap<OsString, DirEntry>>();
+
+    // STEP 3
+    let ctx = ArenaContext(
+        try_files_from_dir(
+            paths
+                .get(OsStr::new("data"))
+                .ok_or_else(|| option_err("Missing data directory from TID directory"))?
+                .path(),
+        )?
+        .into_iter()
+        .map(|data_dat_file| {
+            DataArena::from_bytes(
+                std::fs::read(data_dat_file).wrap_err("Failed to read file from data directory")?,
+            )
+        })
+        .collect::<Result<Vec<_>>>()?,
+    );
+
+    // STEP 4
+    let mut count: usize = 0;
+    try_files_from_dir(
+        paths
+            .get(OsStr::new("ops"))
+            .ok_or_else(|| option_err("Missing ops directory from TID directory"))?
+            .path(),
+    )?
+    // STEP 5
+    .into_iter()
+    .map(|ops_dat_file| {
+        std::fs::read(ops_dat_file)
+            .wrap_err("Failed to read file from ops directory")
+            .and_then(|file_contents| {
+                OpsArena::from_bytes(file_contents)
+                    .wrap_err("Error constructing OpsArena")?
+                    .decode(&ctx)
+                    .wrap_err("Error decoding OpsArena")
+            })
+    })
+    // STEP 6
+    .try_for_each(|x| {
```

Comment:
Rename `x` to `thread_ops`, I think, or a more descriptive name.

Reply:
fixed https://github.com/charmoniumQ/PROBE/pull/20/commits/7d0fa6a25225bb2d0a187c2b741fbb8f80bcc307